common/queue.c

```diff
@@ -0,0 +1,208 @@
+    if (!q || !buffer) {
+        return;
+    }
+    
+    *q = (queue_t *) buffer;
+    (*q)->capacity = capacity;
+    queue_flush(*q);
+}
+
+queue_t *queue_alloc(size_t capacity) {
+    if (capacity == 0) {
+        return NULL;
+    }
+    
+    void *buffer = malloc(queue_calc_size(capacity));
+    if (!buffer) {
+        return NULL;
+    }
+    
+    queue_t *q;
+    queue_init(&q, capacity, buffer);
+    return q;
+}
+
+void queue_destroy(queue_t *q) {
+    free(q);
+}
+
+void queue_flush(queue_t *q) {
+    if (!q) {
+        return;
+    }
+
+    #ifndef HAVE_STDATOMIC_H
+    q->head = 0;
+    q->tail = 0;
+    #else
+    atomic_store_explicit(&q->head, 0, memory_order_relaxed);
+    atomic_store_explicit(&q->tail, 0, memory_order_relaxed);
+    #endif
+}
+
+bool queue_push(queue_t *q, void *item) {
+    if (!q || !item) {
+        return false;
+    }
+    
+    #ifndef HAVE_STDATOMIC_H
+    size_t new_tail = (q->tail + 1) % (q->capacity + 1);
+
+    if (new_tail == q->head) {
+        return false; // Queue full
+    }
+    q->items[q->tail] = item;
+    q->tail = new_tail;
+    #else
+    size_t old_tail = atomic_load_explicit(&q->tail, memory_order_relaxed);
+    size_t new_tail = (old_tail + 1) % (q->capacity + 1);
+
+    // Ensure that the previous head is consumed.
+    if (new_tail == atomic_load_explicit(&q->head, memory_order_acquire)) {
+        return false; // Queue full
+    }
+
+    q->items[old_tail] = item;
+    // Release ensures the write completes before advancing tail
+    atomic_store_explicit(&q->tail, new_tail, memory_order_release);
+    #endif
+    
+    return true;
+}
+
+void *queue_pop_front(queue_t *q, bool peek) {
+    if (queue_is_empty(q)) {
+        return NULL;
+    }
+
+    #ifndef HAVE_STDATOMIC_H
+    void *item = q->items[q->head];
+    if (!peek) {
+        q->head = (q->head + 1) % (q->capacity + 1);
+    }
+    #else
+    size_t old_head = atomic_load_explicit(&q->head, memory_order_relaxed);
+    size_t new_head = (old_head + 1) % (q->capacity + 1);
+    void *item = q->items[old_head];
+
+    if (!peek) {
+        // Release ensures the read completes before advancing head
+        atomic_store_explicit(&q->head, new_head, memory_order_release);
+    }
+    #endif
+    
+    return item;
+}
+
+void *queue_pop_back(queue_t *q, bool peek) {
+    if (queue_is_empty(q)) {
+        return NULL;
+    }
+    
+    #ifndef HAVE_STDATOMIC_H
+    size_t new_tail = (q->tail == 0) ? q->capacity : q->tail - 1;
+    void *item = q->items[new_tail];
+    if (!peek) {
+        q->tail = new_tail;
+    }
+    #else
+    size_t old_tail = atomic_load_explicit(&q->tail, memory_order_relaxed);
+    size_t new_tail = (old_tail == 0) ? q->capacity : old_tail - 1;
+    void *item = q->items[new_tail];
+
+    if (!peek) {
+        // Release ensures the read completes before moving tail
+        atomic_store_explicit(&q->tail, new_tail, memory_order_release);
+    }
+    #endif
+    
+    return item;
+}
```

Comment:
This method isn't thread safe. I would suggest removing it. SPSC fifos can't really index backwards. Especially given the gap between valid items in the FIFO when it's full.

Reply:
Yes I'll remove this, but it does handle the empty slot: If tail is at the empty slot, `tail-1` is the last valid slot.

Reply:
Yes, queue_is_empty() prevents you from looking at the empty space. :)